File: api/services/cpp_adapter.py
"""Python adapter that drives the C++ progbox binary as a subprocess."""

from __future__ import annotations

import json
import os
import re
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path
from typing import Any, Callable

from services import engine_adapter

_VENDOR_DIR = Path(__file__).resolve().parent.parent / "vendor" / "progbox_cpp"
_ANALYSIS_SCRIPT = _VENDOR_DIR / "tools" / "analysis.py"

# Candidate binary paths produced by CMake (platform order matters).
_BINARY_CANDIDATES = [
    _VENDOR_DIR / "build" / "progbox",
    _VENDOR_DIR / "build" / "progbox.exe",
    _VENDOR_DIR / "build" / "Release" / "progbox.exe",
    _VENDOR_DIR / "build" / "Release" / "progbox",
    _VENDOR_DIR / "build" / "x64" / "Release" / "progbox.exe",
    _VENDOR_DIR / "build" / "x64" / "Release" / "progbox",
]


def _normalize_season(raw_season: Any) -> int:
    """Return a valid integer season, defaulting for null or non-numeric values."""
    if type(raw_season) is int:
        return raw_season
    if isinstance(raw_season, str):
        try:
            return int(raw_season)
        except ValueError:
            return 2021
    return 2021


def _resolve_binary() -> Path:
    """Return the C++ binary path, respecting PROGBOX_CPP_BINARY override."""
    override = os.environ.get("PROGBOX_CPP_BINARY")
    if override:
        p = Path(override)
        if not p.is_file():
            raise FileNotFoundError(
                f"PROGBOX_CPP_BINARY={override!r} does not exist"
            )
        return p
    for candidate in _BINARY_CANDIDATES:
        if candidate.is_file():
            return candidate
    raise FileNotFoundError(
        "C++ progbox binary not found. "
        "Run `pnpm build:engine` or set PROGBOX_CPP_BINARY."
    )


def _get_season(export_path: Path) -> int:
    """Extract the season year from the export's gameAttributes."""
    data: dict[str, Any] = json.loads(export_path.read_text(encoding="utf-8"))
    raw_season = data.get("gameAttributes", {}).get("season")
    return _normalize_season(raw_season)


def _filter_export(export_path: Path, teaminfo_path: Path, teams: list[str]) -> dict[str, Any]:
    """Return export JSON with players filtered to the requested teams."""
    data: dict[str, Any] = json.loads(export_path.read_text(encoding="utf-8"))
    if not teams:
        return data
    teaminfo: dict[str, str] = json.loads(teaminfo_path.read_text(encoding="utf-8"))
    allowed_tids = {int(tid) for tid, abbr in teaminfo.items() if abbr in teams}
    data["players"] = [
        p for p in data.get("players", [])
        if p.get("tid") in allowed_tids
    ]
    return data


def _write_input_csv(workspace: Path, export_path: Path, teaminfo_path: Path, teams: list[str]) -> int:
    """Parse export with the vendored exportcleaner and write data/input.csv.

    Returns player_count after filtering.
    """
    ec_mod = engine_adapter.load_exportcleaner_module()
    (workspace / "data").mkdir(parents=True, exist_ok=True)

    # exportcleaner writes data/input.csv relative to cwd; the chdir is scoped
    # and restored in finally. Not thread-safe, but runner.py serialises jobs.
    prev = os.getcwd()
    try:
        os.chdir(workspace)
        df, _ = ec_mod.exportcleaner(
            export_file=str(export_path.resolve()),
            teams=teams,
            teaminfo_file=str(teaminfo_path.resolve()),
        )
    finally:
        os.chdir(prev)

    df.to_csv(workspace / "data" / "input.csv")
    return int(len(df))


def _find_cpp_output_dir(base: Path) -> Path:
    """Find the single timestamped directory the C++ binary created inside *base*."""
    subdirs = [p for p in base.iterdir() if p.is_dir()]
    if len(subdirs) != 1:
        raise RuntimeError(
            f"Expected exactly 1 C++ output directory in {base}, "
            f"found {len(subdirs)}: {[p.name for p in subdirs]}"
        )
    return subdirs[0]


_PROGRESS_RE = re.compile(rb"(\d+)/(\d+)\s*\(\s*(\d+)\s*%\)")


def _stream_cpp_progress(
    proc: subprocess.Popen[bytes],
    callback: Callable[[float, str], None] | None,
) -> None:
    """Read C++ binary stdout and emit progress callbacks."""
    if callback is None:
        proc.stdout.read(4096)
        return

    buf = b""
    last_pct = -1
    is_tty = sys.stdout.isatty()

    while True:
        chunk = proc.stdout.read(4096)
        if not chunk:
            break
        buf += chunk
        lines = re.split(rb"[\r\n]", buf)
        buf = lines[-1]
        complete_lines = lines[:-1]

        for line in complete_lines:
            decoded = line.decode("utf-8", errors="replace").strip()
            if not decoded:
                continue

            if is_tty:
                sys.stdout.write(f"\r{decoded}")
                sys.stdout.flush()

            match = _PROGRESS_RE.search(line)
            if match:
                cpp_pct = int(match.group(3))
                if cpp_pct != last_pct:
                    last_pct = cpp_pct
                    callback(float(cpp_pct), decoded)

    if is_tty and buf:
        print(buf.decode("utf-8", errors="replace").strip())


def run_cpp_simulation(
    export_path: Path,
    teaminfo_path: Path,
    teams: list[str],
    seed: int,
    runs: int,
    n_workers: int,
    canonical_run_dir: Path,
    progress_callback: Callable[[float, str], None] | None = None,
    stage_callback: Callable[[str, str], None] | None = None,
) -> int:
    """Run the full C++ simulation pipeline and return the player count.

    stage_callback fires at post-simulation checkpoints with (stage_name, message):
    "cpp_done", "artifacts_copied", "analyzing", "analysis_done". Lets the runner
    surface progress for the otherwise-silent save + analysis work.
    """
    def _emit_stage(name: str, message: str) -> None:
        if stage_callback is not None:
            stage_callback(name, message)

    binary = _resolve_binary()
    cpp_outputs_base = canonical_run_dir / "_cpp_tmp_outputs"
    cpp_outputs_base.mkdir(parents=True, exist_ok=True)

    try:
        with tempfile.TemporaryDirectory(prefix="progbox_cpp_") as _ws:
            workspace = Path(_ws)

            player_count = _write_input_csv(workspace, export_path, teaminfo_path, teams)

            if teams:
                filtered = _filter_export(export_path, teaminfo_path, teams)
                effective_export = workspace / "export_filtered.json"
                effective_export.write_text(json.dumps(filtered), encoding="utf-8")
            else:
                effective_export = export_path

            season = _normalize_season(_get_season(export_path))
            cmd = [
                str(binary),
                str(effective_export.resolve()),
                str(teaminfo_path.resolve()),
                str(cpp_outputs_base.resolve()),
                "-v", "v41",
                "-r", str(runs),
                "-w", str(n_workers),
                "-s", str(seed),
                "-y", str(season),
            ]
            proc = subprocess.Popen(
                cmd, cwd=str(workspace),
                stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                bufsize=0,
            )
            try:
                _stream_cpp_progress(proc, progress_callback)
                returncode = proc.wait()
            finally:
                if proc.poll() is None:
                    proc.kill()
                    proc.wait()
            if returncode != 0:
                raise RuntimeError(f"C++ binary exited with code {returncode}")

            _emit_stage("cpp_done", "Saving workbook…")

            cpp_run_dir = _find_cpp_output_dir(cpp_outputs_base)
            raw_dst = canonical_run_dir / "raw"
            if raw_dst.exists():
                shutil.rmtree(raw_dst)
            shutil.copytree(cpp_run_dir / "raw", raw_dst)

            cpp_meta_src = cpp_run_dir / "metadata.json"
            if cpp_meta_src.is_file():
                shutil.copy2(cpp_meta_src, canonical_run_dir / "engine_metadata.json")

            _emit_stage("artifacts_copied", "Copied artifacts.")

            # PYTHONUTF8=1 avoids cp1252 crashes on Windows when analysis.py
            # prints box-drawing chars. Non-zero exit is a warning (matches the
            # C++ binary) so tiny filtered runs without enough players for
            # scipy KDE still produce a valid "complete" simulation.
            analysis_env = {**os.environ, "PYTHONUTF8": "1", "PYTHONIOENCODING": "utf-8"}
            _emit_stage("analyzing", "Generating analysis…")
            analysis_result = subprocess.run(
                [sys.executable, str(_ANALYSIS_SCRIPT.resolve()), str(canonical_run_dir.resolve())],
                cwd=str(workspace),
                env=analysis_env,
            )
            if analysis_result.returncode != 0:
                print(
                    f"Warning: tools/analysis.py exited with code "
                    f"{analysis_result.returncode} (analysis_dashboard.html may be incomplete)",
                    flush=True,
                )
            _emit_stage("analysis_done", "Analysis complete.")
    finally:
        shutil.rmtree(cpp_outputs_base, ignore_errors=True)

    return player_count

File: api/src/services/cppAdapter.ts
import { createHash } from "node:crypto";
import { isDeepStrictEqual } from "node:util";
import { normalizeNetInput } from "./netInput.js";
import { spawn } from "node:child_process";
import { once } from "node:events";
import fs from "node:fs";
import fsp from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { vendorCppDir } from "../paths.js";
import {
  DEFAULT_PROGRESSION_VERSION,
  engineScriptId,
  type ProgressionVersion,
} from "../progressionVersions.js";
import { normalizeSeason } from "../utils/normalizeSeason.js";
import { buildInputRows, rowsToCsv } from "./exportCleaner.js";
import { runAnalysis } from "./analysisPython.js";

const VENDOR_DIR = vendorCppDir();
const _BINARY_CANDIDATES = [
  path.join(VENDOR_DIR, "build", process.platform === "win32" ? "progbox.exe" : "progbox"),
  path.join(VENDOR_DIR, "build", "progbox"),
  path.join(VENDOR_DIR, "build", "progbox.exe"),
  path.join(VENDOR_DIR, "build", "Release", "progbox.exe"),
  path.join(VENDOR_DIR, "build", "Release", "progbox"),
  path.join(VENDOR_DIR, "build", "x64", "Release", "progbox.exe"),
  path.join(VENDOR_DIR, "build", "x64", "Release", "progbox"),
];

export function resolveBinary(): string {
  const override = process.env.PROGBOX_CPP_BINARY?.trim();
  if (override) {
    if (!fs.existsSync(override)) {
      throw new Error(`PROGBOX_CPP_BINARY=${JSON.stringify(override)} does not exist`);
    }
    return override;
  }
  for (const c of _BINARY_CANDIDATES) {
    if (fs.existsSync(c)) return c;
  }
  throw new Error("C++ progbox binary not found. Run `pnpm build:engine` or set PROGBOX_CPP_BINARY.");
}

export function getSeason(exportPath: string): number {
  const data = JSON.parse(fs.readFileSync(exportPath, "utf8")) as Record<string, unknown>;
  const raw = (data.gameAttributes as Record<string, unknown> | undefined)?.season;
  return normalizeSeason(raw);
}

export function filterExport(
  exportPath: string,
  teaminfoPath: string,
  teams: string[],
): Record<string, unknown> {
  const data = JSON.parse(fs.readFileSync(exportPath, "utf8")) as Record<string, unknown>;
  if (!teams.length) return data;
  const teaminfo = JSON.parse(fs.readFileSync(teaminfoPath, "utf8")) as Record<string, string>;
  const allowedTids = new Set<number>();
  for (const [tid, abbr] of Object.entries(teaminfo)) {
    if (teams.includes(abbr)) {
      allowedTids.add(parseInt(tid, 10));
    }
  }
  const players = (data.players as unknown[]) ?? [];
  data.players = players.filter((p) => {
    if (!p || typeof p !== "object") return false;
    const tid = (p as Record<string, unknown>).tid;
    return typeof tid === "number" && allowedTids.has(tid);
  });
  return data;
}

export function writeInputCsv(
  workspace: string,
  exportPath: string,
  teaminfoPath: string,
  teams: string[],
): number {
  const exportData = JSON.parse(fs.readFileSync(exportPath, "utf8")) as Record<string, unknown>;
  const teamLookup = JSON.parse(fs.readFileSync(teaminfoPath, "utf8")) as Record<string, string>;
  const rows = buildInputRows(exportData, teamLookup, teams);
  const dataDir = path.join(workspace, "data");
  fs.mkdirSync(dataDir, { recursive: true });
  fs.writeFileSync(path.join(dataDir, "input.csv"), rowsToCsv(rows), "utf8");
  return rows.length;
}

const PROGRESS_RE = /(\d+)\/(\d+)\s*\(\s*(\d+)\s*%\)/;

async function streamCppProgress(
  stdout: NodeJS.ReadableStream | null,
  progressCallback?: (pct: number, message: string) => void,
): Promise<void> {
  if (!stdout) return;
  let buf = "";
  let lastPct = -1;
  for await (const chunk of stdout) {
    const s = typeof chunk === "string" ? chunk : chunk.toString("utf8");
    buf += s;
    const parts = buf.split(/\r?\n/);
    buf = parts.pop() ?? "";
    for (const line of parts) {
      const decoded = line.trim();
      if (!decoded) continue;
      const m = decoded.match(PROGRESS_RE);
      if (m && progressCallback) {
        const cppPct = parseInt(m[3]!, 10);
        if (cppPct !== lastPct) {
          lastPct = cppPct;
          progressCallback(cppPct, decoded);
        }
      }
    }
  }
}

export function findCppOutputDir(base: string): string {
  const subdirs = fs
    .readdirSync(base, { withFileTypes: true })
    .filter((d) => d.isDirectory())
    .map((d) => path.join(base, d.name));
  if (subdirs.length !== 1) {
    throw new Error(
      `Expected exactly 1 C++ output directory in ${base}, found ${subdirs.length}: ${subdirs.map((p) => path.basename(p)).join(", ")}`,
    );
  }
  return subdirs[0]!;
}

export interface RunCppOptions {
  exportPath: string;
  teaminfoPath: string;
  teams: string[];
  seed: number;
  runs: number;
  n_workers: number;
  canonicalRunDir: string;
  progressCallback?: (pct: number, message: string) => void;
  stageCallback?: (stage: string, message: string) => void;
  version?: string;
}

export interface RunCppResult {
  playerCount: number;
  analysisEngine: "python" | "fallback";
}

export async function runCppSimulation(opts: RunCppOptions): Promise<RunCppResult> {
  const emitStage = (name: string, message: string) => {
    opts.stageCallback?.(name, message);
  };

  const binary = resolveBinary();
  const canonicalRunDir = opts.canonicalRunDir;
  const cppOutputsBase = path.join(canonicalRunDir, "_cpp_tmp_outputs");
  fs.mkdirSync(cppOutputsBase, { recursive: true });

  try {
    const workspace = await fsp.mkdtemp(path.join(os.tmpdir(), "progbox_cpp_"));

    try {
      const version = (opts.version as ProgressionVersion | undefined) ?? DEFAULT_PROGRESSION_VERSION;
      let effectiveExport = path.resolve(opts.exportPath);
      let seasonY = getSeason(opts.exportPath);
      let playerCount: number;
      let inputContract: ReturnType<typeof normalizeNetInput>["contract"] | undefined;
      if (version !== "v4.1") {
        const normalized = normalizeNetInput(
          JSON.parse(await fsp.readFile(opts.exportPath, "utf8")),
          JSON.parse(await fsp.readFile(opts.teaminfoPath, "utf8")), opts.teams, version,
        );
        inputContract = normalized.contract;
        seasonY = inputContract.entering_season;
        playerCount = inputContract.target_count;
        if (!playerCount) throw new Error("No eligible NET targets for the selected season and teams");
        effectiveExport = path.join(workspace, "export_normalized.json");
        await fsp.writeFile(effectiveExport, JSON.stringify(normalized.data), "utf8");
      } else {
        playerCount = writeInputCsv(workspace, opts.exportPath, opts.teaminfoPath, opts.teams);
        if (opts.teams.length) {
          effectiveExport = path.join(workspace, "export_filtered.json");
          await fsp.writeFile(effectiveExport, JSON.stringify(filterExport(opts.exportPath, opts.teaminfoPath, opts.teams)), "utf8");
        }
      }

      const cmd = [
        binary,
        effectiveExport,
        path.resolve(opts.teaminfoPath),
        path.resolve(cppOutputsBase),
        "-v",
        engineScriptId(
          (opts.version as ProgressionVersion | undefined) ?? DEFAULT_PROGRESSION_VERSION,
        ),
        "-r",
        String(opts.runs),
        "-w",
        String(opts.n_workers),
        "-s",
        String(opts.seed),
        "-y",
        String(seasonY),
      ];

      // Capture the launch artifact before a concurrent rebuild can replace it
      // while the process is running.
      const binarySha256 = createHash("sha256").update(await fsp.readFile(binary)).digest("hex");
      const proc = spawn(cmd[0]!, cmd.slice(1), {
        cwd: workspace,
        stdio: ["ignore", "pipe", "pipe"],
      });

      const stderrChunks: Buffer[] = [];
      proc.stderr?.on("data", (c: Buffer) => stderrChunks.push(c));

      await Promise.all([
        streamCppProgress(proc.stdout, opts.progressCallback),
        once(proc, "close"),
      ]);
      const code = proc.exitCode ?? 1;
      if (code !== 0) {
        const errText = Buffer.concat(stderrChunks).toString("utf8");
        throw new Error(`C++ binary exited with code ${code}${errText ? `: ${errText}` : ""}`);
      }

      emitStage("cpp_done", "Saving workbook…");

      const cppRunDir = findCppOutputDir(cppOutputsBase);
      const cppMetaSrc = path.join(cppRunDir, "metadata.json");
      let metadata: Record<string, unknown> | undefined;
      if (fs.existsSync(cppMetaSrc)) {
        metadata = JSON.parse(await fsp.readFile(cppMetaSrc, "utf8"));
      }
      if (inputContract) {
        if (!metadata || typeof metadata !== "object" || Array.isArray(metadata)) {
          throw new Error("C++ engine metadata is required for normalized NET input; rebuild the engine");
        }
        if (!isDeepStrictEqual(metadata.input_contract, inputContract)) {
          throw new Error("C++ engine did not acknowledge the requested NET input contract; rebuild the engine");
        }
        const simulation = metadata.simulation as { year?: unknown } | undefined;
        const progression = metadata.progression as { id?: unknown } | undefined;
        if (simulation?.year !== inputContract.entering_season) {
          throw new Error("C++ engine metadata year does not match the NET entering season");
        }
        if (metadata.player_count !== inputContract.target_count) {
          throw new Error("C++ engine metadata player count does not match the NET target count");
        }
        if (progression?.id !== engineScriptId(version)) {
          throw new Error("C++ engine metadata progression does not match the requested NET version");
        }
      }

      // Only publish artifacts after the executing engine confirms the contract.
      const rawDst = path.join(canonicalRunDir, "raw");
      if (fs.existsSync(rawDst)) {
        await fsp.rm(rawDst, { recursive: true, force: true });
      }
      await fsp.cp(path.join(cppRunDir, "raw"), rawDst, { recursive: true });
      if (metadata) {
        if (!inputContract) metadata.input_contract = { id: "legacy-v41" };
        metadata.binary_sha256 = binarySha256;
        await fsp.writeFile(path.join(canonicalRunDir, "engine_metadata.json"), JSON.stringify(metadata, null, 2));
      }

      emitStage("artifacts_copied", "Copied artifacts.");
      emitStage("analyzing", "Generating analysis…");
      const analysisEngine = await runAnalysis(canonicalRunDir);
      emitStage("analysis_done", "Analysis complete.");

      return { playerCount, analysisEngine };
    } finally {
      await fsp.rm(workspace, { recursive: true, force: true }).catch(() => {});
    }
  } finally {
    await fsp.rm(cppOutputsBase, { recursive: true, force: true }).catch(() => {});
  }
}
